types/decimal_test.go

```diff
@@ -105,23 +105,183 @@ func TestDecimal(t *testing.T) {
+		testutil.Equals(t, min, types.DecimalMin)
+
+	})
+
+	t.Run("Cmp", func(t *testing.T) {
+		t.Parallel()
+		one := testutil.Must(types.NewDecimal(1, 0))
+		zero := testutil.Must(types.NewDecimal(0, 0))
+		testutil.Equals(t, one.Cmp(zero), 1)
+		testutil.Equals(t, one.Cmp(one), 0)
+		testutil.Equals(t, zero.Cmp(one), -1)
+	})
+
+	t.Run("NewDecimal", func(t *testing.T) {
+		t.Parallel()
+		tests := []struct {
+			in   int64
+			exp  int
+			want string
+		}{
+			{9223372036854775807, -4, "922337203685477.5807"},
+			{922337203685477580, -3, "922337203685477.58"},
+			{92233720368547758, -2, "922337203685477.58"},
+			{9223372036854775, -1, "922337203685477.5"},
+			{922337203685477, 0, "922337203685477.0"},
+			{-9223372036854775808, -4, "-922337203685477.5808"},
+			{-922337203685477580, -3, "-922337203685477.58"},
+			{-92233720368547758, -2, "-922337203685477.58"},
+			{-9223372036854775, -1, "-922337203685477.5"},
+			{-922337203685477, 0, "-922337203685477.0"},
+			{92233720368547, 1, "922337203685470.0"},
+			{9223372036854, 2, "922337203685400.0"},
+			{922337203685, 3, "922337203685000.0"},
+			{9, 14, "900000000000000.0"},
+		}
+		for _, tt := range tests {
+			t.Run(tt.want, func(t *testing.T) {
+				t.Parallel()
+				d, err := types.NewDecimal(tt.in, tt.exp)
+				testutil.OK(t, err)
+				testutil.Equals(t, d.String(), tt.want)
+			})
+		}
+	})
+
+	t.Run("NewDecimalFromInt", func(t *testing.T) {
+		t.Parallel()
+		tests := []struct {
+			in   int64
+			want string
+		}{
+			{0, "0.0"},
+			{1, "1.0"},
+			{-1, "-1.0"},
+			{922337203685477, "922337203685477.0"},
+			{-922337203685477, "-922337203685477.0"},
+		}
+		for _, tt := range tests {
+			t.Run(tt.want, func(t *testing.T) {
+				t.Parallel()
+				d, err := types.NewDecimalFromInt(tt.in)
+				testutil.OK(t, err)
+				testutil.Equals(t, d.String(), tt.want)
+			})
+		}
+	})
+	t.Run("NewDecimalOverflow", func(t *testing.T) {
+		t.Parallel()
+		tests := []struct {
+			in  int64
+			exp int
+		}{
+			{922337203685477581, -3},
+			{92233720368547759, -2},
+			{9223372036854776, -1},
+			{922337203685478, 0},
+			{92233720368548, 1},
+			{922337203685477581, 2},
+			{10, 14},
+			{1, 15},
+		}
+		for _, tt := range tests {
+			t.Run(fmt.Sprintf("%ve%v", tt.in, tt.exp), func(t *testing.T) {
+				t.Parallel()
+				_, err := types.NewDecimal(tt.in, tt.exp)
+				testutil.ErrorIs(t, err, types.ErrDecimal)
+			})
+		}
+	})
+
+	t.Run("NewDecimalUnderflow", func(t *testing.T) {
+		t.Parallel()
+		tests := []struct {
+			in  int64
+			exp int
+		}{
+			{-922337203685477581, -3},
+			{-92233720368547759, -2},
+			{-9223372036854776, -1},
+			{-922337203685478, 0},
+			{-92233720368548, 1},
+			{-922337203685477581, 2},
+			{-10, 14},
+			{-1, 15},
+		}
+		for _, tt := range tests {
+			t.Run(fmt.Sprintf("%ve%v", tt.in, tt.exp), func(t *testing.T) {
+				t.Parallel()
+				_, err := types.NewDecimal(tt.in, tt.exp)
+				testutil.ErrorIs(t, err, types.ErrDecimal)
```

Comment:
Does ErrDecimal need to be public?  Or is it adequate for it just to emit errors?

Reply:
Let's take care of that in another PR that hides all the errors.